auth.go

```diff
@@ -13,27 +16,54 @@ type AuthenticationResponse struct {
 	Scope        string `json:"scope"`
 }
 
-// Authenticate sends an authentication request, and returns a newly constructed client object
+var cache = ttlcache.New[string, string]()
+
+// Authenticate sends an authentication request when a token is not available in the cache, and returns a newly constructed client object
 func Authenticate(username, password, apiEndpoint string) (*Client, error) {
 	infolog.Printf("Start authentication into '%s' using '%s'", apiEndpoint, LoginUrl)
+	userAgent := ConstructUserAgentString()
+	cachedToken := getCachedAccessToken(username, apiEndpoint)
+	if len(cachedToken) > 0 {
+		infolog.Printf("Returning auth token from cache")
+		return &Client{Username: username, Password: password, ApiEndpoint: apiEndpoint, UserAgent: userAgent}, nil
+	} else {
+		values := map[string]string{"username": username, "password": password}
+		jsonData, err := json.Marshal(values)
+		if err != nil {
+			return nil, ConstructNestedError("error during json marshalling", err)
+		}
+		resp, err, _ := request(context.TODO(), "", "POST", apiEndpoint+LoginUrl, userAgent, nil, string(jsonData))
+		if err != nil {
+			return nil, ConstructNestedError("authentication request failed", err)
+		}
 
-	values := map[string]string{"username": username, "password": password}
-	jsonData, err := json.Marshal(values)
-	if err != nil {
-		return nil, ConstructNestedError("error during json marshalling", err)
-	}
+		var authResp AuthenticationResponse
+		if err = jsonStrictUnmarshall(resp, &authResp); err != nil {
+			return nil, ConstructNestedError("failed to unmarshal authentication response with error", err)
+		}
 
-	userAgent := ConstructUserAgentString()
-	resp, err := request(context.TODO(), "", "POST", apiEndpoint+LoginUrl, userAgent, nil, string(jsonData))
-	if err != nil {
-		return nil, ConstructNestedError("authentication request failed", err)
+		infolog.Printf("Authentication was successful")
+		cache.Set(getCacheKey(username, apiEndpoint), authResp.AccessToken, time.Duration(authResp.ExpiresIn)*time.Millisecond)
+		return &Client{Username: username, Password: password, ApiEndpoint: apiEndpoint, UserAgent: userAgent}, nil
 	}
+}
+
+// getCacheKey calculates an access token key using the username and the apiEndpoint provided
+func getCacheKey(username, apiEndpoint string) string {
+	return username + apiEndpoint
```

Comment:
We could hash but I'm not sure it would bring benefits (all the data is available in the dsn so I don't think there is a point hiding the data in memory). Also, it makes things easier when trying to debug (eg, if a test fails we can put a breaking and check the content of the cache if needed)